test/jdk/jdk/crac/CracEngineOptionsTest.java

```diff
@@ -0,0 +1,180 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import org.junit.Test;
+import org.junit.BeforeClass;
+import static org.junit.Assert.*;
+
+import jdk.test.lib.Platform;
+import jdk.test.lib.Utils;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+import java.nio.file.Path;
+import java.util.*;
+
+/*
+* @test
+* @summary Testing CRaCEngine and CRaCEngineOptions VM options.
+* @library /test/lib
+* @build CracEngineOptionsTest
+* @run junit/othervm CracEngineOptionsTest
+*/
+public class CracEngineOptionsTest {
+    @BeforeClass
+    public static void checkCriu() {
+        final boolean hasCriu = Path.of(Utils.TEST_JDK, "lib", "criuengine").toFile().exists();
+        assertEquals("CRIU exists iff we are on Linux", Platform.isLinux(), hasCriu);
+    }
+
+    @Test
+    public void test_default() throws Exception {
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(
+                "-XX:CRaCCheckpointTo=cr",
+                "-version");
+        OutputAnalyzer out = new OutputAnalyzer(pb.start());
+        out.shouldHaveExitValue(0);
+    }
+
+    @Test
+    public void test_engines() throws Exception {
+        test("sim");
+        test("simengine");
+        test("pause");
+        test("pauseengine");
+        if (Platform.isLinux()) {
+            test("criu");
+            test("criuengine");
+        }
+
+        final String absolute = Platform.isWindows() ?
+            Path.of(Utils.TEST_JDK, "bin", "simengine.exe").toString() :
+            Path.of(Utils.TEST_JDK, "lib", "simengine").toString();
+        test(absolute);
+
+        test("unknown", Collections.emptyList(), 1, "Cannot find CRaC engine unknown");
+        test("simengine,--arg", Collections.emptyList(), 1, "Cannot find CRaC engine simengine,--arg");
+        test("one two", Collections.emptyList(), 1, "Cannot find CRaC engine one two");
+        test("", Collections.emptyList(), 1, "CRaCEngine must not be empty");
+    }
+
+    @Test
+    public void test_options() throws Exception {
+        test("simengine", "");
+        test("simengine", "image_location=cr", 0,
+                "Internal CRaC engine option provided, skipping: image_location");
+        if (Platform.isLinux()) {
+            test("criuengine", "keep_running=true,args=-v -v -v -v,keep_running=false", 0,
+                    "CRaC engine option: 'keep_running' = 'true'",
+                    "CRaC engine option: 'args' = '-v -v -v -v'",
+                    "CRaC engine option: 'keep_running' = 'false'");
```

Comment:
I think that providing the same option twice should result in warning, at least.

Reply:
The engine is not interpreting JVM options; engine receives its programmatic configuration and has to deal with that. JVM defines the user API and we can decide that repeated use of the same argument is prohibited (as that would be more likely an user error).

Reply:
I mean that an engine can interpret reconfiguration of the same key not as overwriting of the previous value but somehow differently, e.g. as concatenation in my example. Since the VM is just passing the keys through I would leave the interpretation up to the engine. But it's not a strong opinion, I can move the warning to JVM if you still think it would be better.

Reply:
After thinking a bit more, yes, let's move the warning to JVM to simplify the engine implementation. We can always remove the warning if we find a sensible use case for repeating keys.